### PR TITLE
bass-tube database dev backup

### PR DESCRIPTION
**Pull request by sf-chris**
- fetches fixture data database
- serializes data fetched from database
- builds "insert_into" queries used to construct database subset

*Originally created on GitLab by Codestein (https://gitlab.com/sportsflare/real-time/bass-tube/-/merge_requests/95) on 2023-03-12T21:53:34.443Z*

### CODE REVIEW

**Comment by sf-chris** [comment]
Generally I believe it's to do with the pagination of the returned data, in this case it looks like it's only returning id's which are gte to the cursor value.

if cursor != 0 makes no sense and is clearly wrong here

*Originally commented on GitLab by Codestein on 2023-03-14T22:19:04.915Z*

**Comment by sf-chris** [comment]
Make changes according to your comments. Unsure whether the deserialization function is really required as it's only converting it into a tuple. Have left it for now to match provided specs.

Most filtering options removed. Initial function now accepts a list of SportAlias types, which will gather fixtures from specific sports.

*Originally commented on GitLab by Codestein on 2023-03-14T22:11:53.978Z*

**Comment by sf-chris** [comment]
I really think none of this should live in the storage repo. 

Along with Steffan's comment.

Please make this all self contained. I really don't think it needs to live in the storage repo.

Remember the goal is to pull this into different code bases that all use async pg. So if we're needing to sprinkle functions around the code base then its going to be a pain to try and migrate.

Just make a self contained module that has like 3 functions:

1) Run a query to get shit from postgres
2) Serailize that into a format that can be written to a file (literally just a tuple)
3) Write it to a file

*Originally commented on GitLab by sadboy on 2023-03-13T21:55:46.291Z*

**Comment by sf-chris** [comment]
I think this fucntion is over general. Just make it specific please. I want to get fixtures that match this description.

Please just write one function that does a query that gives you an entire list of what you want. That function can run a query per title given a list of titles.

*Originally commented on GitLab by sadboy on 2023-03-13T21:55:46.016Z*

**Comment by sf-chris** [comment]
These aren't needed. Do you know why a cursor is used?

*Originally commented on GitLab by sadboy on 2023-03-13T21:55:45.925Z*

**Comment by sf-chris** [comment]
Columns don'e need to be set every time. just grab them once off the table

So just import the table and grab the columns.

*Originally commented on GitLab by sadboy on 2023-03-13T21:55:45.766Z*

**Comment by sf-chris** [comment]
You don't need to hold a connection for the entire duration of this function. just use it to grab what you need and release it

*Originally commented on GitLab by sadboy on 2023-03-13T21:55:45.678Z*

**Comment by sf-chris** [comment]
agree here

*Originally commented on GitLab by sadboy on 2023-03-13T21:55:45.434Z*

**Comment by sf-chris** [comment]
Starting to look great, had a few comments to add

*Originally commented on GitLab by steffansafey on 2023-03-13T02:56:50.279Z*

**Comment by sf-chris** [comment]
Does this `IndexError` get thrown trying to access `serialized_fixtures[0]`?

We can probably avoid having to catch this error by first checking whether `any(serialized_fixtures) == True`

*Originally commented on GitLab by steffansafey on 2023-03-13T02:56:50.179Z*

**Comment by sf-chris** [comment]
I reckon we should make a new directory for this new process (like `./bass_tube/backup/`)

Reason being the `./bass_tube/storage` module already exists and has a different purpose to this code (it being the data access layer shared between various processes)

*Originally commented on GitLab by steffansafey on 2023-03-13T02:56:49.838Z*

**Comment by sf-chris** [comment]
In terms of building the query, why are you having a new insert statement each time. You can do an insert statement which takes a list of tuples:

Example from PG docs

INSERT INTO films (code, title, did, date_prod, kind) VALUES
    ('B6717', 'Tampopo', 110, '1985-02-10', 'Comedy'),
    ('HG120', 'The Dinner Game', 140, DEFAULT, 'Comedy');

*Originally commented on GitLab by sadboy on 2023-03-12T23:01:14.059Z*

**Comment by sf-chris** [comment]
More i think about it we can just delete this funciton and format it in the build_backup_queries

*Originally commented on GitLab by sadboy on 2023-03-12T23:00:33.404Z*

**Comment by sf-chris** [comment]
This is outstanding code for a first attempt:

- I really like __main__.py using the app structure
- Good work on reusing existing functions and correctly using existing patterns (__main__, app.py, etc)
- backup.py is well named
-

*Originally commented on GitLab by sadboy on 2023-03-12T22:59:37.418Z*

**Comment by sf-chris** [comment]
First impression is that there's a lot of loops here. You should always be asking yourself if you *really really really* need to be looping here. And I suspect not.

below you're looping through the tupled_fixtures, and the tupled_fixtures are created by looping through the fixtures and appending a tuple.

1. you can do that in the loop below `for fixtures in tupled_fixtures`
2. I don't think you need to be so disciplined with your types. You can just pass some string formatting here "f({values})" and then inside that is simply values = ", ".join(f[c] for c in column_names) or equivalent
3. Moreover it makes me wonder before you even get here, why doesn't your "sesrialize_fixtures" function just do this in the first place. so building them is as simple as looping through and creating the insert statement with the values?

*Originally commented on GitLab by sadboy on 2023-03-12T22:59:37.119Z*

**Comment by sf-chris** [comment]
Not your fault, but Matt gave this the type annotation of "str" and you're passing a None so Pycharm is complaining about it. If you can be bothered, it might be worth just double checking the signatures of these functions are correct.

tldr: you guys need a better IDE smh

*Originally commented on GitLab by sadboy on 2023-03-12T22:59:36.973Z*

**Comment by sf-chris** [comment]
@steffansafey also please review

*Originally commented on GitLab by Codestein on 2023-03-12T21:53:49.217Z*
